src/backend/distributed/transaction/backend_data.c

```diff
@@ -0,0 +1,480 @@
+
+	CheckCitusVersion(ERROR);
+
+	/* check to see if caller supports us returning a tuplestore */
+	if (returnSetInfo == NULL || !IsA(returnSetInfo, ReturnSetInfo))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("set-valued function called in context " \
+						"that cannot accept a set")));
+	}
+
+	if (!(returnSetInfo->allowedModes & SFRM_Materialize))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("materialize mode required, but it is not " \
+						"allowed in this context")));
+	}
+
+	/* build a tuple descriptor for our result type */
+	if (get_call_result_type(fcinfo, NULL, &tupleDescriptor) != TYPEFUNC_COMPOSITE)
+	{
+		elog(ERROR, "return type must be a row type");
+	}
+
+	perQueryContext = returnSetInfo->econtext->ecxt_per_query_memory;
+
+	oldContext = MemoryContextSwitchTo(perQueryContext);
+
+	tupleStore = tuplestore_begin_heap(true, false, work_mem);
+	returnSetInfo->returnMode = SFRM_Materialize;
+	returnSetInfo->setResult = tupleStore;
+	returnSetInfo->setDesc = tupleDescriptor;
+
+	MemoryContextSwitchTo(oldContext);
+
+	/*
+	 * We don't want to initialize memory while spinlock is held so we
+	 * prefer to do it here. This initialization is done only for the first
+	 * row.
+	 */
+	memset(values, 0, sizeof(values));
+	memset(isNulls, false, sizeof(isNulls));
+
+	/* we're reading all the backend data, take a lock to prevent concurrent additions */
+	LWLockAcquire(AddinShmemInitLock, LW_SHARED);
+
+	for (backendIndex = 0; backendIndex < MaxBackends; ++backendIndex)
+	{
+		BackendData *currentBackend =
+			&backendManagementShmemData->backends[backendIndex];
+
+		SpinLockAcquire(&currentBackend->mutex);
+
+		/* we're only interested in active backends */
+		if (currentBackend->transactionId.transactionNumber == 0)
+		{
+			SpinLockRelease(&currentBackend->mutex);
+			continue;
+		}
+
+		values[0] = ObjectIdGetDatum(currentBackend->databaseId);
+		values[1] = Int32GetDatum(ProcGlobal->allProcs[backendIndex].pid);
+		values[2] = Int32GetDatum(currentBackend->transactionId.initiatorNodeIdentifier);
+		values[3] = UInt64GetDatum(currentBackend->transactionId.transactionNumber);
+		values[4] = TimestampTzGetDatum(currentBackend->transactionId.timestamp);
+
+		SpinLockRelease(&currentBackend->mutex);
+
+		tuplestore_putvalues(tupleStore, tupleDescriptor, values, isNulls);
+
+		/*
+		 * We don't want to initialize memory while spinlock is held so we
+		 * prefer to do it here. This initialization is done for the rows
+		 * starting from the second one.
+		 */
+		memset(values, 0, sizeof(values));
+		memset(isNulls, false, sizeof(isNulls));
+	}
+
+	LWLockRelease(AddinShmemInitLock);
+
+	/* clean up and return the tuplestore */
+	tuplestore_donestoring(tupleStore);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * InitializeBackendManagement requests the necessary shared memory
+ * from Postgres and sets up the shared memory startup hook.
+ */
+void
+InitializeBackendManagement(void)
```

Comment:
Taste (you choose): I'd name these `*BackendData` rather than `*BackendManagement`.